File: src/LongestPalindromicSubstring.java
/**
 Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.

 Example:

 Input: "babad"

 Output: "bab"

 Note: "aba" is also a valid answer.
 Example:

 Input: "cbbd"

 Output: "bb"

 */
public class LongestPalindromicSubstring {
    public static void main(String[] args) {
        LongestPalindromicSubstring l = new LongestPalindromicSubstring();

        Printer.printResultComparsion(l.longestPalindrome("babad"), "bab");
    }

    public String longestPalindrome(String s) {
        char[] arr = s.toCharArray();
        int n = arr.length, i, maxLen = 0;
        String res = "";

        for (i = 0; i < n; i++) {
            // check string two characters bigger than current max;
            if (isPalindrome(s, i - maxLen - 1, i)) {
                res = s.substring(i - maxLen - 1, i + 1);
                maxLen = maxLen + 2;
            }

            // check string one character bigger than current max;
            if (isPalindrome(s, i - maxLen, i)) {
                res = s.substring(i - maxLen, i + 1);
                maxLen = maxLen + 1;
            }
        }

        return res;
    }

    public boolean isPalindrome(String s, int start, int end) {
        if (start < 0 || end >= s.length()) {
            return false;
        }

        while (start < end) {
            if (s.charAt(start) == s.charAt(end)) {
                start++;
                end--;
            } else {
                return false;
            }
        }
        return true;
    }

    public String longestPalindromeUsingDp(String s) {
        int n = s.length();

        String result = null;

        boolean[][] dp = new boolean[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = i; j >= 0; j--) {
                if (s.charAt(i) == s.charAt(j)) {
                    if ((i - j) <= 2 || dp[i - 1][j + 1]) {
                        dp[i][j] = true;

                        if (result == null ||
                                (i - j + 1 > result.length())) {
                            result = s.substring(j, i+1);
                        }
                    }
                }
            }
        }

        return result;
    }}

File: python/longest_palindromic_substring.py
'''

200. Longest Palindromic Substring
Given a string S, find the longest palindromic substring in S.
You may assume that the maximum length of S is 1000, and there exists one unique longest palindromic substring.

Example
Given the string = "abcdzdcab", return "cdzdc".

Challenge
O(n2) time is acceptable. Can you do it in O(n) time.

'''

class Solution:
    """
    @param s: input string
    @return: the longest palindromic substring
    """
    def longestPalindrome(self, s):
        # write your code here
        len_s, start, end = len(s), 0, 0
        len_m = 0
        for i in range(1, 2 * len_s):
            if i % 2 == 0:
                left = i//2
                right = left
            else:
                left = i//2 - 1
                right = left + 1
            while left >= 0 and right < len_s and s[left] == s[right]:
                left -= 1
                right += 1
            left += 1
            right -= 1
            len_t = right - left + 1
            if len_t > len_m:
                len_m = len_t
                start = left
                end = right
        return s[start:end+1]